### PR TITLE
Fix reading fixed_len_byte_array of length 9-16 in parquet

### PR DESCRIPTION
**Pull request by raunaqmorarka**
If the extra bytes are all zeros or all -1 for negative values,
then the remaining bytes can be decoded as a short decimal.

Currently incorrect results are returned when a negative value is 
stored in fixed_len_byte_array of size 9-16 and read as a short decimal.

### CODE REVIEW

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java [hunk 1]
```diff
@@ -147,4 +162,31 @@ else if (columnDescriptor.getPrimitiveType().getPrimitiveTypeName() == INT64) {
             }
         }
     }
+
+    private void checkBytesFitInShortDecimal(byte[] bytes, int startOffset, Type trinoType)
+    {
+        if ((bytes[startOffset] & 0x80) != 0) {
```
Comment (skrzypo987) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java:168]: Extract that condition and name it something like `negative`

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java [hunk 2]
```diff
@@ -264,4 +269,9 @@ public static Slice getLongDecimalValue(byte[] bytes)
         BigInteger value = new BigInteger(bytes);
         return unscaledDecimal(value);
     }
+
+    public static boolean isNegative(byte[] bytes, int offset)
```
Comment (martint) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:273]: This refactoring should go in a separate commit.

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java [hunk 3]
```diff
@@ -264,4 +269,9 @@ public static Slice getLongDecimalValue(byte[] bytes)
...
+    {
+        return (bytes[offset] & 0x80) != 0;
```
Comment (martint) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:275]: This seems like a convoluted way to say `bytes[offset] < 0`

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java [hunk 4]
```diff
@@ -147,4 +163,31 @@ else if (columnDescriptor.getPrimitiveType().getPrimitiveTypeName() == INT64) {
             }
         }
     }
+
+    private void checkBytesFitInShortDecimal(byte[] bytes, int startOffset, Type trinoType)
+    {
+        if (isNegative(bytes, startOffset)) {
```
Comment (martint) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java:169]: That `startOffset` name is misleading. That's where the check stops, not where it starts.

Also, why do we check `isNegative` on `startOffset` instead of the more obvious `bytes[0]`?

Comment (martint) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java:169]: You can restructure this method to:
* check if the first value is 0 or -1
* loop to check if all the other values are equal to the first one

I think it will make the implementation shorter and easier to follow.

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java (hunk 3 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java]: yes, i've simplified this in a separate commit now

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java (hunk 4 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java]: Simplified this method now.
Also renamed startOffset to endOffset here.
I'm checking `bytes[endOffset]` instead of `bytes[0]` so that we show an error when trying to read an encoding where the 8th byte from the end is negative but the remaining upper bytes are 0.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java [hunk 5]
```diff
@@ -147,4 +162,21 @@ else if (columnDescriptor.getPrimitiveType().getPrimitiveTypeName() == INT64) {
             }
         }
     }
+
+    private void checkBytesFitInShortDecimal(byte[] bytes, int endOffset, Type trinoType)
+    {
+        byte expectedValue = 0;
+        if (bytes[endOffset] < 0) {
```
Comment (martint) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java:169]: There's a trick to do this in a branchless manner:

`bytes[endOffset] >> 7`

If the value is is positive, it produces a 0. If it's negative, it produces a -1 by extending the sign to all the lower bits.

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ShortDecimalColumnReader.java]: nice, updated to that now

**Review by martint — APPROVED** [review]
